### PR TITLE
Use variables in pipeline

### LINKED ISSUE #936: Global Environment Variables
### Clear and concise description of the problem

I've notice that there is no way to declare a given environment variable to be global to all defined steps.

In `GitLab CI`, for instance, there is a "global" `variables` key that fulfills that purpose.

For instance, take a look at this manifest:

```yaml
default:
  cache:
    key: $CI_COMMIT_REF_SLUG
    paths:
      - $CI_PROJECT_DIR/.m2/repository/
  image: docker.io/library/eclipse-temurin:8-jdk-focal

stages:
  - compile

variables:
  MAVEN_CONFIG: --batch-mode --no-transfer-progress
  MAVEN_OPTS: -Dmaven.repo.local=$CI_PROJECT_DIR/.m2/repository/

compile:
  stage: compile
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_MERGE_REQUEST_ID
  script:
    - ./mvnw compile $MAVEN_EXTRA_OPTS
  variables:
    MAVEN_EXTRA_OPTS: -DskipTests=true
```

Here `MAVEN_CONFIG` and `MAVEN_OPTS` apply to all steps. As far as I can tell, currently there is no way to accomplish this in `Woodpecker CI`.

### Suggested solution

An `environment` key at the root level.

> The `environment` key at the step level would have to account for overrides, in case the same environment variable is declared in both.

### Alternative

N/A

### Additional context

N/A

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by 6543**
use yaml aliases (https://yaml.org/spec/1.2.2/#3222-anchors-and-aliases) to have pipeline `variables`

[read more](https://www.educative.io/blog/advanced-yaml-syntax-cheatsheet)

TODO:
 - [x] apply to own pipeline config
 - [x] update pipeline schema
 - [x] document it

---
close  #936

### CODE REVIEW

**Comment by lafriks** [comment]
Can environment variables be used in variable values?

**Comment by 6543** [comment]
of course, this variables are a yaml feature and so all things we parse or extend come afterwards

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    image: golang:1.18
+    command: build
+```
+
+just use a new section called **variables**:
+
+```diff
++variables:
++  - &golang_image 'golang:1.18'
++
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:23]:
```suggestion

```

**Comment by 6543** [comment]
this pull just document it and extend the schema to not complain

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# Advanced YAML tricks
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:1]: Is there a better term than `tricks`? Tricks sounds like some unofficial way of doing it IMO.

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+-    image: golang:1.18
++    image: *golang_image
+     command: go test ./...
+   build:
+-    image: golang:1.18
++    image: *golang_image
+     command: build
+```
+
+## Overrides and Extensions
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:35]:
```suggestion
## Example of YAML override and extension
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+pipelines:
+  develop:
+    name: Build and test
+    image: some-plugin
+    settings: *some-plugin-settings
+    when:
+      branch: develop
+
+  master:
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:53]:
```suggestion
  main
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    name: Build and test
+    image: some-plugin
+    settings:
+      <<: *some-plugin-settings
+      try: false #override
+      ongoing: false #extension
+    when:
+      branch: master
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:61]:
```suggestion
      branch: main
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    when:
+      branch: develop
+
+  master:
+    name: Build and test
+    image: some-plugin
+    settings:
+      <<: *some-plugin-settings
+      try: false #override
+      ongoing: false #extension
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:59]:
```suggestion
      ongoing: false # adding a new value to `some-plugin-settings`
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    settings: *some-plugin-settings
+    when:
+      branch: develop
+
+  master:
+    name: Build and test
+    image: some-plugin
+    settings:
+      <<: *some-plugin-settings
+      try: false #override
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:58]:
```suggestion
      try: false # replacing original value from `some-plugin-settings`
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md (hunk 2 above)
Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md:1]: "usage" ?

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md:1]: Or syntax?

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md:1]: just push for one and I'll change it ...

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md:1]:
```suggestion
# Advanced YAML syntax
```

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  build:
+    image: golang:1.18
+    command: build
+```
+
+Just use a new section called **variables**:
+
+```diff
++variables:
++  - &golang_image 'golang:1.18'
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:22]:
```suggestion
+  - &golang_image 'golang:1.18'

```

Ups, I just wanted to remove the `+` in the previous suggestion.

File: docs/docs/20-usage/35-advanced-yaml-tricks.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
+# Advanced YAML tricks
```
Comment (anbraten) [inline docs/docs/20-usage/35-advanced-yaml-tricks.md:1]: Could you link this page from the pipeline syntax list somehow like we did it for `secrets`.

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: I dont understand

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: Adding a new paragraph to https://woodpecker-ci.org/docs/usage/pipeline-syntax on the top-level like `services` with a content similar to this: 

> Woodpecker provides to use [YAML anchors & aliases}(https://yaml.org/spec/1.2.2/#3222-anchors-and-aliases) in the pipeline configuration. These can be used to write DRY configs by using variables.
> 
> For more details check the [Advanced YAML syntax docs](https://woodpecker-ci.org/docs/usage/advanced-yaml-syntax).

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: I dont think we should link it there ...

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: Why not wouldn't it be incomplete somehow if we don't link it? 🤔

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: well it's no specific pipeline config thing ... this document yaml spec behaviour

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: I get your point, but it now somehow has a specific position in the config with `variables` now. In addition I think users would search how to use variables quite often as seen in the issues, so linking it here as well could be helpful.

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: @anbraten can you push to this branch to make the specific docs changes?

Comment (anbraten) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: I think the branch is protected, but I can create a PR against it.

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-tricks.md]: @anbraten **https://github.com/6543-forks/woodpecker/invitations**

**Review by anbraten — APPROVED** [review]

**Comment by svdHero** [comment]
@6543 just to clarify: This is not released, yet, is it? I couldn't find this piece of information about anchors & aliases in the official documentation.
Will it only be released with version 1.0.0?

File: docs/docs/20-usage/35-advanced-yaml-syntax.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+-    image: golang:1.18
++    image: *golang_image
+     command: go test ./...
+   build:
+-    image: golang:1.18
++    image: *golang_image
+     command: build
+```
+
+## Example of YAML override and extension
```
Comment (svdHero) [inline docs/docs/20-usage/35-advanced-yaml-syntax.md:35]: @6543 shouldn't this second paragraph also be displayed [here](https://woodpecker-ci.org/docs/next/usage/advanced-yaml-syntax)? I only see the first paragraph. But maybe I misunderstood the purpose of the "next documentation".

Comment (6543) [reply docs/docs/20-usage/35-advanced-yaml-syntax.md:35]: @svdHero it's not working so it go hiden until fixed

**Pull request merged**